### DISCUSSION TITLE
2-10-2 Internal PID set Input

### DISCUSSION THREAD

**Discussion by HansGem** [artisan-roaster-scope/artisan#1492] [General]
1. I was hoping to get PID to follow an alternative device available in input dropdown  but no SV comes through and no effective control of slider. Specifically, with an air roaster I am wanting to follow <(R2-RB2)+(Y2-B2)> with SV=0

Rationale: currently PID Input is BT, and result sees BT 1degree c under most of the way and the after first crack clime above BT.

3. Also  - can anyone explain or send me where I can find re (a) "dervitaive filter" and (b) 'p on error' and 'p on input?'

**Comment by MAKOMO** [comment]
It is quite hard to decipher your discussion point. Maybe something got lost as I see a point 1. and 3. but no point 2.

What does 2-10-2 stand for? What does it mean that "no SV comes through" and what exactly you want to follow and what does the "<(R2-RB2)+(Y2-B2)> with SV=0" stand for?

The "derivative filter" option activates a first-order filter on the D component. In the following you find information on the 'P on error' vs 'P on input' of PID design: http://brettbeauregard.com/blog/2017/06/introducing-proportional-on-measurement/

**Reply by HansGem** [reply]
2-10-2 = version of Artisan I am using. 

Thanks Marco for the reference and explanation. The explaination of P-Error vs P-Measurement will se me experimenting with the latter.  I am not sure what a first order filter means. 

I was trying to be concise in the previous. Here is  detailed explanation. 

Currently the PID, input BT,  p on error, controlls gas flow to the burner (in an air roaster), and goes well after drying,  all good up to first crack, the BT  tracking around about 1 degree below the background BT . After first crack the ROR drops and then takes off, as anticipated as the classical drop and flick. However the PID doesn't seem to want to correct the sudden rise, leaving it going up, and if I increase P any more than what I had calibrated, the temp just overshoots,  bouncing around, up and down too much. 

So, I wanted to experiment with a PID input based on _the sum of_ how much rate of rise and bean temp deviate from their respective backgrounds, instead of just BT error.  So, I have configured a virtual device as (R2-RB2)+(Y2-B2) - showing up as an appropriate curve. 

This virtual device shows nicely in the  drop down in the 'inbuilt' Artisan PID,  allowing  me to select this virtual device instead of BT. In this scenario  I have sought to set the SV manually to "0" as  the target for the virtual device. However, on the runs i have done to this point based on P-error, I just get completely erratic behaviour. I will try P on error, but still miffed as to D setting and filter. 

Querie: If I set Input to any option like a virtual device, and I leave background on, will the calculations be based on the difference between the background trace of that variable and the current - ie SV based on background trace? 

In the attached, the green tags  are manual adjustments in parallel or post PID running: airflow adjustments, (slowing=hotter) and the red second burner adjustments. 

![1124 - 2024-03-03_1503 (1) Ethiopia Yirgacheffe Idido G1 11 5  09_43min  217 2C](https://github.com/artisan-roaster-scope/artisan/assets/62494802/a9628c13-eefe-45ee-8878-7bdd62b59482)

many thanks

HansR

**Comment by MAKOMO** [comment]
> 2-10-2 = version of Artisan I am using.
> 
> Thanks Marco for the reference and explanation. The explaination of P-Error vs P-Measurement will se me experimenting with the latter. I am not sure what a first order filter means.

https://www.google.de/search?q=first+order+filter

> 
> I was trying to be concise in the previous. Here is detailed explanation.
> 
> Currently the PID, input BT, p on error, controlls gas flow to the burner (in an air roaster), and goes well after drying, all good up to first crack, the BT tracking around about 1 degree below the background BT . After first crack the ROR drops and then takes off, as anticipated as the classical drop and flick. However the PID doesn't seem to want to correct the sudden rise, leaving it going up, and if I increase P any more than what I had calibrated, the temp just overshoots, bouncing around, up and down too much.

Sounds like you need to tune your p-i-d settings a little better.

> 
> So, I wanted to experiment with a PID input based on _the sum of_ how much rate of rise and bean temp deviate from their respective backgrounds, instead of just BT error. So, I have configured a virtual device as (R2-RB2)+(Y2-B2) - showing up as an appropriate curve.

What unit would such a reading have? The `(R2-RB2)` part is in C/min while the `(Y2-B2)` part is in C. Sounds like apples with pears...

> 
> This virtual device shows nicely in the drop down in the 'inbuilt' Artisan PID, allowing me to select this virtual device instead of BT. In this scenario I have sought to set the SV manually to "0" as the target for the virtual device. However, on the runs i have done to this point based on P-error, I just get completely erratic behaviour. I will try P on error, but still miffed as to D setting and filter.
> 
> Querie: If I set Input to any option like a virtual device, and I leave background on, will the calculations be based on the difference between the background trace of that variable and the current - ie SV based on background trace?

Yes. If you choose BT as input, with SV on `background `, the SV will taken from the background profiles BT at the current sampling time (modulo lookahead). If you choose 0xT1 (the first curve of the first extra device) as input, the SV will taken from the first curve of the backgrounds first extra device.

> 
> In the attached, the green tags are manual adjustments in parallel or post PID running: airflow adjustments, (slowing=hotter) and the red second burner adjustments.
> 
> ![1124 - 2024-03-03_1503 (1) Ethiopia Yirgacheffe Idido G1 11 5 09_43min 217 2C](https://private-user-images.githubusercontent.com/62494802/310364993-a9628c13-eefe-45ee-8878-7bdd62b59482.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDk3MDQ1NzEsIm5iZiI6MTcwOTcwNDI3MSwicGF0aCI6Ii82MjQ5NDgwMi8zMTAzNjQ5OTMtYTk2MjhjMTMtZWVmZS00NWVlLTg4NzgtN2JkZDYyYjU5NDgyLmpwZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMDYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzA2VDA1NTExMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWNjNDk5YmMzZjlhMmUxMGNjNmMxMzcxZTM3OTM4YzlkZGI5ZTBkZjZiOWUwMTNhZTBkZjU5NjQ5MGViYWU1NTYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.RnopTqqUjSNBoTSqfsiwQxVsx_le_mFC7GDOMO3IeoE)
> 
> many thanks
> 
> HansR

**Reply by HansGem** [reply]
It will be a couple of days (weekend) before I get to playing and running more trials. 
In the meantime, results yesterday using P-Measurement look promising - in that scenario if I understand correctly (thanks for the ref!), we are not introducing a background based SV to calculate "error"  but using the measure entered from the virtual device.   
(I am a psychometrician trying to make sense of engineering 🤔😕 ! )
"Apples and pears"  ie the IVs being different metrics doesn't matter to my mind-  in this scenario both measures we are using are what in stats we call scaled derivatives, ie  we are simply looking at difference scores as determined by the equations, not absolute C and absolute delta. 
Now, consider if only using say BT as input, and its travelling on line (Y2-B2=0) then PID adjusting gas setting (D) would stay put. 
But (R2-RB2) in this case however is useful as it would _predict_ where one is going, a metric telling us if BT is going to be staying on track, or moving up or down, and to what degree. It should thus be profitable to include it some way (perhaps appropriately weighted) to impact gas adjustment, ie "D."
This is as much as I can work out from my limited understanding and the limited information I can glean of the computational processes, ie  IVs  determining the P-Measurement  DVs (Dependent Variables) ie P, I, and D. 
Hope this makes some sort of sense.

**Reply by HansGem** [reply]
Just to report nice progress: On an air roaster, pid controlling gas slider, I am now pretty happy with: pid set to 10~.2~20, p on error, mode = manual, SV thus set to zero, and device added: ((Y2-B2)+(R2-RB2))0*.5. and this selected as input.
I can boost at the first crack drop by temporarily raising SV at the first crack drop and flick.
Otherwise R2-RB2 nicely adjusts before Y2-B2 starts to come off line . All good. Many thanks.
